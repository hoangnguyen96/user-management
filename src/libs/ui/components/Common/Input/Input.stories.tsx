import type { Meta, StoryObj } from "@storybook/react";
import { Input } from ".";
import { Box } from "@mui/material";

const InputBase = ({ ...rest }) => (
  <Box display="flex">
    <Input {...rest} />
  </Box>
);

const meta: Meta<typeof Input> = {
  component: InputBase,
};

export default meta;
type Story = StoryObj<typeof Input>;

export const Default: Story = {
  args: {
    placeholder: "Input...",
    variant: "outlined",
  },
};

export const Error: Story = {
  args: {
    placeholder: "Input...",
    isInvalid: true,
  },
};
