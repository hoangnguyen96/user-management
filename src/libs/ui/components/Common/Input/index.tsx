import { forwardRef, memo } from "react";
import {
  Stack,
  TextField,
  TextFieldProps,
  TextFieldVariants,
  Typography,
} from "@mui/material";
import { themeDefault } from "../../../themes";

interface CustomTextFieldProps extends Omit<TextFieldProps, "variant"> {
  variant?: TextFieldVariants;
  errorMessage?: string;
  isInvalid?: boolean;
}

export const Input = memo(
  forwardRef<HTMLDivElement, CustomTextFieldProps>(
    ({ variant, errorMessage, isInvalid, type = "text", ...rest }, ref) => {
      return (
        <Stack ref={ref} mb={errorMessage ? 0 : "20px"}>
          <TextField
            type={type}
            variant={variant}
            sx={{
              "& .MuiOutlinedInput-root": {
                "& fieldset": {
                  borderColor:
                    errorMessage || isInvalid
                      ? themeDefault().palette.error.main
                      : "#b5b7C0",
                },
                "&:hover fieldset": {
                  borderColor:
                    errorMessage || isInvalid
                      ? themeDefault().palette.error.main
                      : themeDefault().palette.primary.main,
                },
                "&.Mui-focused fieldset": {
                  borderColor:
                    errorMessage || isInvalid
                      ? themeDefault().palette.error.main
                      : themeDefault().palette.primary.main,
                },
              },
            }}
            {...rest}
          />
          {errorMessage && (
            <Typography
              variant="caption"
              fontSize="12px"
              ml="10px"
              color={themeDefault().palette.error.main}
            >
              {errorMessage}
            </Typography>
          )}
        </Stack>
      );
    }
  )
);
