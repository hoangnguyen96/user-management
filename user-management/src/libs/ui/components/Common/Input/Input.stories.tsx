import type { Meta, StoryObj } from "@storybook/react";
import { Input } from ".";
import { Box } from "@mui/material";

const InputBase = ({ ...rest }) => (
  <Box display="flex">
    <Input {...rest} />
  </Box>
);

const meta: Meta<typeof Input> = {
  component: InputBase,
};

export default meta;
type Story = StoryObj<typeof Input>;

export const Normal: Story = {
  args: {
    placeholder: "Input...",
    variant: "normal",
  },
};

export const Warning: Story = {
  args: {
    placeholder: "Input...",
    variant: "warning",
  },
};

export const Error: Story = {
  args: {
    placeholder: "Input...",
    variant: "error",
  },
};
