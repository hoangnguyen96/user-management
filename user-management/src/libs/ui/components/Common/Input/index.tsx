import { forwardRef, memo } from "react";
import {
  Stack,
  SxProps,
  TextField,
  TextFieldProps,
  Theme,
  Typography,
} from "@mui/material";
import { themeDefault } from "../../../themes";

// Define custom variant types
type CustomVariant = "error" | "warning" | "normal";

interface CustomTextFieldProps extends Omit<TextFieldProps, "variant"> {
  variant?: CustomVariant;
  errorMessage?: string;
  isInvalid?: boolean;
}

const customStyles: Record<CustomVariant, SxProps<Theme>> = {
  error: {
    "& .MuiOutlinedInput-root": {
      "& fieldset": {
        borderColor: "red",
      },
      "&:hover fieldset": {
        borderColor: "darkred",
      },
      "&.Mui-focused fieldset": {
        borderColor: "red",
      },
    },
    "& .MuiInputLabel-root": {
      color: "red",
    },
    "& .MuiInputLabel-root.Mui-focused": {
      color: "red",
    },
    "& .MuiFormHelperText-root": {
      color: "red",
    },
  },
  warning: {
    "& .MuiOutlinedInput-root": {
      "& fieldset": {
        borderColor: "orange",
      },
      "&:hover fieldset": {
        borderColor: "darkorange",
      },
      "&.Mui-focused fieldset": {
        borderColor: "orange",
      },
    },
    "& .MuiInputLabel-root": {
      color: "orange",
    },
    "& .MuiInputLabel-root.Mui-focused": {
      color: "orange",
    },
    "& .MuiFormHelperText-root": {
      color: "orange",
    },
  },
  normal: {
    "& .MuiOutlinedInput-root": {
      "& fieldset": {
        borderColor: "gray",
      },
      "&:hover fieldset": {
        borderColor: "darkgray",
      },
      "&.Mui-focused fieldset": {
        borderColor: "gray",
      },
    },
    "& .MuiInputLabel-root": {
      color: "gray",
    },
    "& .MuiInputLabel-root.Mui-focused": {
      color: "gray",
    },
    "& .MuiFormHelperText-root": {
      color: "gray",
    },
  },
};

export const Input = memo(
  forwardRef<HTMLDivElement, CustomTextFieldProps>(
    ({ variant, errorMessage, isInvalid, type = "text", ...rest }, ref) => {
      return (
        <Stack ref={ref} mb={errorMessage ? 0 : "20px"}>
          <TextField
            type={type}
            sx={{
              "& .MuiOutlinedInput-root": {
                "& fieldset": {
                  borderColor:
                    errorMessage || isInvalid
                      ? themeDefault.palette.error.main
                      : "#b5b7C0",
                },
                "&:hover fieldset": {
                  borderColor:
                    errorMessage || isInvalid
                      ? themeDefault.palette.error.main
                      : themeDefault.palette.primary.main,
                },
                "&.Mui-focused fieldset": {
                  borderColor:
                    errorMessage || isInvalid
                      ? themeDefault.palette.error.main
                      : themeDefault.palette.primary.main,
                },
              },
              ...(variant ? customStyles[variant] : { variant: "normal" }),
            }}
            {...rest}
          />
          {errorMessage && (
            <Typography
              variant="caption"
              fontSize="12px"
              ml="10px"
              color={themeDefault.palette.error.main}
            >
              {errorMessage}
            </Typography>
          )}
        </Stack>
      );
    }
  )
);
