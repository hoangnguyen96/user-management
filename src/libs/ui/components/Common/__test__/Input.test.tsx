import { render } from "@testing-library/react";

// Component
import { Input } from "../Input";

describe("Input component", () => {
  it("Should render snapshot correctly", () => {
    expect(render(<Input placeholder="Input..." />)).toMatchSnapshot();
  });
});
