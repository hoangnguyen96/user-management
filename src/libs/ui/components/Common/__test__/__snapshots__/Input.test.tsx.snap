// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Should render snapshot correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-aubdti-MuiStack-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1fpxxrg-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              placeholder="Input..."
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-ihdtdm"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-aubdti-MuiStack-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1fpxxrg-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            placeholder="Input..."
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
